flocker/node/agents/blockdevice.py

```diff
@@ -112,6 +127,201 @@ class UnattachedVolume(VolumeException):
     u"A block-device-backed dataset is being created.",
 )
 
+DESTROY_BLOCK_DEVICE_DATASET = ActionType(
+    u"agent:blockdevice:destroy",
+    [VOLUME],
+    [],
+    u"A block-device-backed dataset is being destroyed.",
+)
+
+UNMOUNT_BLOCK_DEVICE = ActionType(
+    u"agent:blockdevice:unmount",
+    [VOLUME],
+    [],
+    u"A block-device-backed dataset is being unmounted.",
+)
+
+DETACH_VOLUME = ActionType(
+    u"agent:blockdevice:detach_volume",
+    [VOLUME],
+    [],
+    u"The volume for a block-device-backed dataset is being detached."
+)
+
+DESTROY_VOLUME = ActionType(
+    u"agent:blockdevice:destroy_volume",
+    [VOLUME],
+    [],
+    u"The volume for a block-device-backed dataset is being destroyed."
+)
+
+
+def _logged_statechange(cls):
+    """
+    Decorate an ``IStateChange.run`` implementation with partially automatic
+    logging.
+
+    :param cls: An ``IStateChange`` implementation which also has an
+        ``_action`` attribute giving an Eliot action that should be used to log
+        its ``run`` method.
+
+    :return: ``cls``, mutated so that its ``run`` method is automatically run
+        in the context of its ``_action``.
+    """
+    original_run = cls.run
+    # Work-around https://twistedmatrix.com/trac/ticket/7832
+    try:
+        original_run.__name__ = original_run.methodName
+    except AttributeError:
+        pass
+
+    @wraps(original_run)
+    def run(self, deployer):
+        with self._action:
+            # IStateChange.run nominally returns a Deferred.  Hook it up to the
+            # action properly.  Do this as part of FLOC-1549 or maybe earlier.
+            return original_run(self, deployer)
+
+    cls.run = run
+    return cls
+
+
+class BlockDeviceVolume(PRecord):
+    """
+    A block device that may be attached to a host.
+
+    :ivar unicode blockdevice_id: The unique identifier of the block device.
```

Comment:
It needs to be unique across the cluster.  This is for values like an EBS volume id, not the local, OS-level identifier for the device.  It needs to be cluster-wide unique otherwise you won't know which block device you're trying to attach/detach/destroy.